File: apps/web-vite/src/effects/definitions/lens-distortion.ts
import type { EffectDefinition } from "@/effects/types";

export const LENS_DISTORTION_SHADER = "lens-distortion";

export const lensDistortionEffectDefinition: EffectDefinition = {
	type: "lens-distortion",
	name: "Lens Distortion",
	keywords: ["lens", "distortion", "barrel", "pincushion", "fisheye", "warp"],
	params: [
		{ key: "distortion", label: "Distortion", type: "number", default: 0, min: -100, max: 100, step: 1 },
		{ key: "zoom", label: "Zoom", type: "number", default: 0, min: 0, max: 100, step: 1 },
	],
	renderer: {
		passes: [{
			shader: LENS_DISTORTION_SHADER,
			uniforms: ({ effectParams }) => ({
				u_scalar0: (Number(effectParams.distortion) || 0) / 100,
				u_scalar1: (Number(effectParams.zoom) || 0) / 100,
			}),
		}],
	},
};

File: rust/crates/effects/src/definitions/lens_distortion.rs
use std::collections::HashMap;

use crate::{EffectCategory, EffectDefinition, EffectParam, EffectUniformBuffer, UniformValue};

fn pack_uniforms(
    uniforms: &HashMap<String, UniformValue>,
    width: u32,
    height: u32,
) -> EffectUniformBuffer {
    let s0 = uniforms
        .get("u_scalar0")
        .and_then(|v| match v {
            UniformValue::Number(n) => Some(*n),
            _ => None,
        })
        .unwrap_or(0.0);
    let s1 = uniforms
        .get("u_scalar1")
        .and_then(|v| match v {
            UniformValue::Number(n) => Some(*n),
            _ => None,
        })
        .unwrap_or(0.0);

    EffectUniformBuffer {
        resolution: [width as f32, height as f32],
        direction: [0.0, 0.0],
        scalars: [s0, s1, 0.0, 0.0],
        _cg_expansion: [0u8; 480],
    }
}

pub const LENS_DISTORTION: EffectDefinition = EffectDefinition {
    id: "lens-distortion",
    name: "Lens Distortion",
    category: EffectCategory::Distort,
    shader_source: include_str!("../shaders/lens_distortion.wgsl"),
    entry_point: "fragment_main",
    uniform_size: std::mem::size_of::<EffectUniformBuffer>(),
    params: &[
        EffectParam {
            name: "u_scalar0",
            label: "Distortion",
            default: 0.0,
            min: -1.0,
            max: 1.0,
            step: 0.01,
            animatable: true,
        },
        EffectParam {
            name: "u_scalar1",
            label: "Zoom",
            default: 0.0,
            min: 0.0,
            max: 1.0,
            step: 0.01,
            animatable: true,
        },
    ],
    pack_uniforms,
};
